### ISSUE TITLE
Deletion of array item (containing watch) reliably crashes JS and corrupts form data

### ISSUE THREAD

**Issue by JElchison** [json-editor/json-editor#812]
### General information

To the entire community, THANKS for such a wonderful project!  I have used json-editor in production for years, with relatively few issues.

My userbase recently reported data corruption after deleting an array item (which itself contains a watch reference).

- json-editor version: 2.3.0

### Steps to reproduce the behavior

https://json-editor.github.io/json-editor/ can reproduce the issue using the following schema:

```javascript
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "properties": {
    "students": {
      "type": "array",
      "format": "tabs",
      "items": {
        "type": "object",
        "id": "student_arr_item",
        "properties": {
          "name": {
            "type": "string"
          },
          "sessions": {
            "type": "array",
            "format": "table",
            "items": {
              "type": "object",
              "properties": {
                "student_name": {
                  "type": "string",
                  "watch": {
                    "watch_student_name": "student_arr_item.name"
                  },
                  "template": "{{ watch_student_name }}"
                },
                "minutes": {
                  "type": "integer"
                }
              }
            }
          }
        }
      }
    }
  }
}
```

...and the following form data:

```javascript
{
  "students": [
    {
      "name": "AAA",
      "sessions": [
        {
          "student_name": "AAA",
          "minutes": 15
        },
        {
          "student_name": "AAA",
          "minutes": 15
        }
      ]
    },
    {
      "name": "BBB",
      "sessions": [
        {
          "student_name": "BBB",
          "minutes": 20
        }
      ]
    },
    {
      "name": "CCC",
      "sessions": [
        {
          "student_name": "CCC",
          "minutes": 10
        }
      ]
    }
  ]
}
```

Steps to reproduce:
1. Visit https://json-editor.github.io/json-editor/
2. Paste in the above schema.  Click "Update Schema" button.
3. Paste in the above form data.  Click "Update Form" button.
4. Open browser's Javascript console.
5. With "Item 1" selected by default, click "Delete item" button once, which will attempt to delete Item 1 (AAA).  The deletion will succeed.  Items BBB and CCC correctly remain.
6. With "Item 1" selected by default, click "Delete item" button once more, which will attempt to delete Item 1 (BBB).

### Expected behavior

Correct deletion of array item.  There should be 1 item remaining (CCC).  No data corruption.

### Actual behavior

Failed deletion of array item.  There are 2 items remaining (CCC and CCC).  Data is corrupted.

Observe following JS error on console:
```
jsoneditor.min.js:18 Uncaught TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at object.refreshValue (jsoneditor.min.js:18)
    at object.onChildEditorChange (jsoneditor.min.js:18)
    at y.change (jsoneditor.min.js:18)
    at y.onChange (jsoneditor.min.js:18)
    at y.setValue (jsoneditor.min.js:18)
    at y.onWatchedFieldChange (jsoneditor.min.js:18)
    at Array.watch_listener (jsoneditor.min.js:18)
    at K.notifyWatchers (jsoneditor.min.js:18)
    at y.notify (jsoneditor.min.js:18)
    at y.onChange (jsoneditor.min.js:18)
    at y.setValue (jsoneditor.min.js:18)
    at jsoneditor.min.js:18
    at Array.forEach (<anonymous>)
    at object.setValue (jsoneditor.min.js:18)
    at jsoneditor.min.js:18
    at Array.forEach (<anonymous>)
    at w.setValue (jsoneditor.min.js:18)
    at HTMLButtonElement.<anonymous> (jsoneditor.min.js:18)
```
This sequence is 100% repeatable (all 50 of ~50 attempts).

Environment:
* Chrome Version 85.0.4183.83 (Official Build) (64-bit)
* Ubuntu 20.04.1 LTS running kernel `Linux xxxxxxxx 5.4.0-45-generic #49-Ubuntu SMP Wed Aug 26 13:38:52 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`

Based on the traceback, it seems as if the list of watchers may not be updated properly upon array item deletion.

I am happy to do any requested legwork.  But as a Javascript novice, I could use some help pointing me in the right direction.

Thanks in advance!

**Comment by schmunk42** [comment]
Confirmed.

@germanbisurgi Could you create a test (or maybe two with a set of 5 items).

**Comment by germanbisurgi** [comment]
I can reproduce the error following the gives steps. After that i tried to manually input the equivalent data in the editors an work perfectly. I presume there is a problem with setValue() + watchers (?), maybe "notify"

**Comment by JElchison** [comment]
Hi friends, I've written a [test](https://github.com/json-editor/json-editor/compare/master...JElchison:dev-issue-gh-812) for this issue, and I've started down the path of more intense debugging.

I'm fairly certain that the issue is that the `jsoneditor.watchlist` is never updated when an Array item is removed.

_After_ the first item deletion in the above test case, I observe that the watchlist looks like this:
![image](https://user-images.githubusercontent.com/1757771/114301321-48126580-9a92-11eb-9b25-197276ff28ac.png)

I would have expected it to look more like this _after_ the first deletion:
![image](https://user-images.githubusercontent.com/1757771/114301379-86a82000-9a92-11eb-9b21-b1f8ac626230.png)

Instead, no matter how many items are deleted from the array, the watchlist remains unchanged from its first value (always 3 items).

With the second item deletion, then, the model is inconsistent since the first item in the watchlist no longer has 2 watchers.  This leads to a crash.

I tested this theory by setting `jsoneditor.watchlist = null` inside of `_createDeleteButton`, and the test I wrote started passing.  However, this is not a valid fix--it just confirmed my theory.

I could use some direction on how to implement a real fix.  Since watchlist is an Object itself (with item names as strings), it appears unnatural to modify it directly.  It seems to me that I need to cause the watchlist to be regenerated from the new contents of the Array (i.e., with the appropriate item removed), but I'm unable to find a concise way to do that.

Any tips?  Thanks in advance!
